src/main.cpp

```diff
@@ -0,0 +1,122 @@
+//
+// Module for generating meshes from point clouds
+//
+#include <vector>
+#include <pcl/point_types.h>
+#include <pcl/io/pcd_io.h>
+#include <pcl/io/vtk_io.h>
+#include <pcl/io/vtk_lib_io.h>
+#include <pcl/search/kdtree.h>
+#include <pcl/search/organized.h>
+#include <pcl/features/normal_3d_omp.h>
+#include <pcl/surface/gp3.h>
+#include <pcl/filters/passthrough.h>
+#include <pcl/filters/radius_outlier_removal.h>
+#include <pcl/surface/poisson.h>
+#include <ros/ros.h>
+#include <pcl_ros/point_cloud.h>
+#include "mesh_generator/MeshPointCloud.h"
+#include <pcl_msgs/PolygonMesh.h>
+
+
+typedef mesh_generator::MeshPointCloud::Request MeshPointCloudRequest;
+typedef mesh_generator::MeshPointCloud::Response MeshPointCloudResponse;
+typedef pcl::PointCloud<pcl::PointXYZ> PointCloud;
+typedef pcl::PointCloud<pcl::Normal> NormalCloud;
+
+/**
+ * Estimates the normals for a point cloud using PCL
+ *
+ * @param point_cloud The input point cloud that the normals are generated for
+ * @return Returns a NormalCloud::Ptr with the generated normals
+ */
+NormalCloud::Ptr estimate_normals(PointCloud::Ptr point_cloud)
+{
+    ROS_INFO("Estimating normals");
+
+    // Declare PCL objects needed to perform normal estimation
+    pcl::NormalEstimationOMP<pcl::PointXYZ, pcl::Normal> normal_estimation;
+    NormalCloud::Ptr normals(new NormalCloud);
+    pcl::search::KdTree<pcl::PointXYZ>::Ptr search_tree(new pcl::search::KdTree<pcl::PointXYZ>);
+
+    // Set input parameters for normal estimation
+    search_tree->setInputCloud(point_cloud);
+    normal_estimation.setInputCloud(point_cloud);
+    normal_estimation.setSearchMethod(search_tree);
+    normal_estimation.setKSearch(20);
+
+    // Perform normal estimation algorithm
+    normal_estimation.compute(*normals);
+
+    // Reverse the direction of all normals
+    for (size_t i = 0; i < normals->size(); ++i) {
+        normals->points[i].normal_x *= -1;
+        normals->points[i].normal_y *= -1;
+        normals->points[i].normal_z *= -1;
+    }
+
+    return normals;
+}
+
+/**
+ * Uses poisson surface reconstruction algorithm to generate a 3D-mesh from a point cloud
+ * with normals
+ *
+ * @param point_cloud The point cloud (including normals) that is used to generate the mesh
+ * @param mesh The resulting 3D-mesh is saved to mesh
```

Comment:
Var lite otydligt förklarat, det menas att resultatet av meshningen sparas till objektet som heter mesh (som skickas in som argument via en referens så man kommer åt resultatet även utanför funktionen).